server/grpc_service.go

```diff
@@ -1969,18 +1969,23 @@ func (s *GrpcServer) WatchGlobalConfig(req *pdpb.WatchGlobalConfigRequest, serve
 	// If the revision is compacted, will meet required revision has been compacted error.
 	// - If required revision < CompactRevision, we need to reload all configs to avoid losing data.
 	// - If required revision >= CompactRevision, just keep watching.
-	watchChan := s.client.Watch(ctx, configPath, clientv3.WithPrefix(), clientv3.WithRev(revision))
+	// Use WithPrevKV() to get the previous key-value pair when get Delete Event.
+	watchChan := s.client.Watch(ctx, configPath, clientv3.WithPrefix(), clientv3.WithRev(revision), clientv3.WithPrevKV())
 	for {
 		select {
 		case <-ctx.Done():
 			return nil
 		case res := <-watchChan:
-			if revision < res.CompactRevision {
-				if err := server.Send(&pdpb.WatchGlobalConfigResponse{
-					Revision: res.CompactRevision,
-					Header: s.wrapErrorToHeader(pdpb.ErrorType_DATA_COMPACTED,
-						fmt.Sprintf("required watch revision: %d is smaller than current compact/min revision. %d", revision, res.CompactRevision)),
-				}); err != nil {
+			if res.Err() != nil {
+				var resp pdpb.WatchGlobalConfigResponse
+				if revision < res.CompactRevision {
+					resp.Header = s.wrapErrorToHeader(pdpb.ErrorType_DATA_COMPACTED,
```

Comment:
directly return `error`.

Reply:
The errors in Err() are that need to be retried so return error at the end of the if

Reply:
The client is likely to need special logic for DataCompacted error, if we just return err here, can the client distinguish between different errors